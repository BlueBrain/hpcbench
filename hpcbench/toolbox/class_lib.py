"""Provide metaclass to allow top-hierarchy classes
to access its subclasses
"""

__all__ = [
    'ClassRegistrar'
]


class ClassLibrary(object):
    """ClassRegistrar modify top-level classes so that they extend
    this class.

    This class should not used directly
    """
    @classmethod
    def get_subclass(cls, name):
        """Get a subclass

        :param name: Name of the class to retrieve
        """
        return cls.SUB_CLASSES[name]  # pylint: disable=no-member

    @classmethod
    def get_subclasses(cls):
        """Get list of available classes
        :return: list of class names
        :rtype: list of string
        """
        for clazz in cls.SUB_CLASSES:  # pylint: disable=no-member
            yield clazz

    @classmethod
    def register_subclass(cls, clazz):
        """Register a subclass

        :param clazz: Class to register
        """
        name = getattr(clazz, 'name', clazz.__name__)
        classes = cls.SUB_CLASSES  # pylint: disable=no-member
        if name in classes:
            raise Exception('class %s is already registered' % name)
        classes[name] = clazz


class ClassRegistrar(type):
    """Metaclass that can be specified on top-hierarchy classes
    to track their subclasses.

    Usage:

    >>> class Foo(object):
            __metaclass__ = ClassRegistrar
    >>>    class Bar(Foo): pass
    >>> Foo.get_subclass('Bar')
    <class '__main__.Foo'>

    A custom name can be specified:

    >>> class EmbarassingClassName(Foo):
            name = "awesome-class"
    >>> Foo.get_subclass('awesome-class')
    <class '__main__.EmbarassingClassName'>
    """
    def __new__(mcs, name, bases, attrs):
        if not bases or bases == (object,):
            attrs['SUB_CLASSES'] = {}
            bases = (ClassLibrary,) + bases
        cls = type.__new__(mcs, name, bases, attrs)
        cls.register_subclass(cls)
        return cls
