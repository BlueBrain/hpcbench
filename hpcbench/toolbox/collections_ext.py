"""Extra collections utilities
"""
import collections
import copy
import errno
import os
import os.path as osp
import sys

import six
import yaml


class nameddict(dict):  # pragma pylint: disable=invalid-name
    """ Provides dictionary whose keys are accessible via the property
    syntax: `obj.key`
    """
    def __init__(self, *args, **kwargs):
        dict.__init__(self, *args, **kwargs)
        self.__dict__ = self
        self.__namify(self.__dict__)

    @classmethod
    def __namify(cls, a_dict):
        for key in a_dict.keys():
            if isinstance(a_dict[key], dict):
                a_dict[key] = nameddict(a_dict[key])

    def __setitem__(self, key, value):
        if isinstance(value, dict):
            value = nameddict(value)
        super(nameddict, self).__setitem__(key, value)

    def __setattr__(self, key, value):
        if key != '__dict__' and isinstance(value, dict):
            value = nameddict(value)
        super(nameddict, self).__setattr__(key, value)

    def __deepcopy__(self, memo):
        cls = self.__class__
        content = dict()
        for key, value in self.items():
            content[key] = copy.deepcopy(value, memo)
        result = cls.__new__(cls)
        result.__init__(content)
        return result


class Configuration(nameddict):
    """nameddict reflecting a YAML file
    """
    @classmethod
    def from_file(cls, path):
        """Create a ``Configuration`` from a file

        :param path: path to YAML file
        :return: new configuration
        :rtype: ``Configuration``
        """
        if path == '-':
            return Configuration(yaml.safe_load(sys.stdin))
        if not osp.exists(path) and not osp.isabs(path):
            path = osp.join(osp.dirname(osp.abspath(__file__)), path)
        with open(path, 'r') as istr:
            return Configuration(yaml.safe_load(istr))

    @classmethod
    def from_env(cls, envvars, default, default_config):
        """Create a ``Configuration``

        :param envvars: list of environment variable name where to look
        for a YAML filename
        :param default: default path to YAML file if not found in ``envvars``
        :param default_config: Default ``Configuration`` if not found
        in above methods.
        :return: new configuration
        :rtype: ``Configuration``
        """
        try:
            if isinstance(envvars, six.string_types):
                envvars = [envvars]
            config_file = default
            for envvar in envvars:
                envvalue = os.getenv(envvar)
                if envvalue is not None:
                    config_file = envvalue
                    break
            config = Configuration.from_file(config_file)
        except IOError as exc:
            if exc.errno in [errno.ENOENT, errno.ENOTDIR]:
                config = default_config
            else:
                raise
        return config


def flatten_dict(dic, parent_key='', sep='.'):
    """Flatten sub-keys of a dictionary
    """
    items = []
    for key, value in dic.items():
        new_key = parent_key + sep + key if parent_key else key
        if isinstance(value, collections.MutableMapping):
            items.extend(flatten_dict(value, new_key, sep=sep).items())
        else:
            items.append((new_key, value))
    return dict(items)


def dict_merge(dct, merge_dct):
    """ Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
    updating only top-level keys, dict_merge recurses down into dicts nested
    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
    ``dct``.
    :param dct: dict onto which the merge is executed
    :param merge_dct: dct merged into dct
    :return: None
    """
    for key in merge_dct.keys():
        if (key in dct and isinstance(dct[key], dict)
                and isinstance(merge_dct[key], collections.Mapping)):
            dict_merge(dct[key], merge_dct[key])
        else:
            dct[key] = merge_dct[key]


def dict_map_kv(obj, func):
    if isinstance(obj, collections.Mapping):
        return {
            func(k): dict_map_kv(v, func)
            for k, v in six.iteritems(obj)
        }
    elif isinstance(obj, list):
        return [dict_map_kv(e, func) for e in obj]
    else:
        return func(obj)
