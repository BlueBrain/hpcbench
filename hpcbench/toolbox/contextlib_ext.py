"""Additional methods usable in with-context
"""
import contextlib
import os
import os.path as osp
import shutil
import sys
import tempfile
import timeit

import six


@contextlib.contextmanager
def capture_stdout():
    """Intercept standard output in a with-context
    :return: cStringIO instance

    >>> with capture_stdout() as stdout:
            ...
        print stdout.getvalue()
    """
    stdout = sys.stdout
    sys.stdout = six.moves.cStringIO()
    try:
        yield sys.stdout
    finally:
        sys.stdout = stdout


@contextlib.contextmanager
def write_wherever(file_name=None):
    writer = open(file_name, 'w') if file_name is not None else sys.stdout
    yield writer
    if file_name is not None:
        writer.close()


@contextlib.contextmanager
def pushd(path, mkdir=True, cleanup=False):
    """Change current working directory in a with-context
    :param mkdir: If True, then directory is created if it does not exist
    :param cleanup: If True and no pre-existing directory, the directory is
    cleaned up at the end
    """
    cwd = os.getcwd()
    exists = osp.exists(path)
    if mkdir and not exists:
        os.makedirs(path)
    os.chdir(path)
    try:
        yield path
    finally:
        os.chdir(cwd)
        if not exists and cleanup:
            # NB: should we be checking for rmtree.avoids_symlink_attacks ?
            shutil.rmtree(path)


class Timer(object):  # pylint: disable=too-few-public-methods
    """Object usable in with-context to time it.
    """
    def __init__(self):
        self.start = None
        self.end = None
        self.timer = timeit.Timer()

    def __call__(self):
        return self.timer.timer()

    def __enter__(self):
        self.start = self()
        return self

    def __exit__(self, exc_type, exc_value, exc_traceback):
        self.end = self()

    @property
    def elapsed(self):
        """
        :return: duration in seconds spent in the context.
        :rtype: float
        """
        if self.end is None:
            return self() - self.end
        return self.end - self.start


@contextlib.contextmanager
def mkdtemp(*args, **kwargs):
    """Create a temporary directory in a with-context

    keyword remove: Remove the directory when leaving the
    context if True. Default is True.
    other keywords arguments are given to the tempfile.mkdtemp
    function.
    """
    remove = kwargs.pop('remove', True)
    path = tempfile.mkdtemp(*args, **kwargs)
    try:
        yield path
    finally:
        if remove:
            shutil.rmtree(path)
