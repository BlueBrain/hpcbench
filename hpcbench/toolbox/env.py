import os
from string import Template


class TemplateWithDefaults(Template):
    idpattern = r'(?:\#)?[_a-z][_a-z0-9]*(?::?-[^}]*)?'

    # Modified from python2.7/string.py
    def substitute(self, mapping):
        # Helper function for .sub()
        def convert(mo):
            # Check the most common path first.
            named = mo.group('named') or mo.group('braced')
            if named is not None:
                if named.startswith('#'):
                    var = named[1:]
                    return str(len(mapping.get(var)))
                if ':-' in named:
                    var, _, default = named.partition(':-')
                    return mapping.get(var) or default
                if '-' in named:
                    var, _, default = named.partition('-')
                    return mapping.get(var, default)
                val = mapping[named]
                return '%s' % (val,)
            if mo.group('escaped') is not None:
                return self.delimiter
            if mo.group('invalid') is not None:
                self._invalid(mo)
            raise ValueError('Unrecognized named group in pattern',
                             self.pattern)
        return self.pattern.sub(convert, self.template)


def expandvars(s, vars=None):
    """Perform variable substitution on the given string

    Supported syntax:
    * $VARIABLE
    * ${VARIABLE}
    * ${#VARIABLE}
    * ${VARIABLE:-default}

    :param s: message to expand
    :type s: str

    :param vars: dictionary of variables. Default is ``os.environ``
    :type vars: dict

    :return: expanded string
    :rtype: str
    """
    tpl = TemplateWithDefaults(s)
    return tpl.substitute(vars or os.environ)
