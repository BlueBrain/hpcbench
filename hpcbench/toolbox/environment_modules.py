import os
from subprocess import PIPE, Popen

from .process import find_executable


class Module:
    MODULECMD = find_executable('modulecmd', required=False)

    @classmethod
    def load(cls, *args):
        cls.python('load', *args)

    @classmethod
    def unload(cls, *args):
        cls.python('load', *args)

    @classmethod
    def purge(cls, *args):
        cls.python('purge', *args)

    @classmethod
    def python(cls, *args):
        command = [cls.MODULECMD, 'python'] + list(args)
        with open(os.devnull, 'w') as devnull:
            proc = Popen(command, shell=False, stdout=PIPE, stderr=devnull)
            stdout, _ = proc.communicate()
            exec(stdout)
