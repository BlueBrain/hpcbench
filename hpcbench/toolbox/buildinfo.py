"""Extract build information from executables
"""

import json
try:
    from json import JSONDecodeError as JSONDcdError
except ImportError:
    JSONDcdError = ValueError
import logging
import subprocess

from hpcbench.toolbox.collections_ext import byteify
from hpcbench.toolbox.contextlib_ext import (
    mkdtemp,
    pushd,
)


LOGGER = logging.getLogger('hpcbench')

OBJCOPY = 'objcopy'
DUMP_SECTION = '--dump-section'
ELF_SECTION = 'build_info'
BUILDINFO_FILE = 'buildinfo.json'


def extract_build_info(exe_path, elf_section=ELF_SECTION):
    """Extracts the build information from a given executable.

    The build information is expected to be in json format, which is parsed
    and returned as a dictionary.
    If no build information is found an empty dictionary is returned.

    This assumes binutils 2.25 to work.

    Args:
        exe_path (str): The full path to the executable to be examined

    Returns:
        dict: A dictionary of the extracted information.
    """
    build_info = {}
    with mkdtemp() as tempd, pushd(tempd):
        errno = subprocess.call([OBJCOPY,
                                 DUMP_SECTION,
                                 "{secn}={ofile}".format(secn=elf_section,
                                                         ofile=BUILDINFO_FILE),
                                 exe_path])
        if errno:  # just return the empty dict
            LOGGER.warning('objcopy failed with errno %s', errno)
            return build_info

        with open(BUILDINFO_FILE) as build_info_f:
            try:
                build_info = json.load(build_info_f, object_hook=byteify)
            except JSONDcdError as jsde:
                LOGGER.warning('benchmark executable build is not valid json:')
                LOGGER.warning(jsde.msg)
                LOGGER.warning('build info section content:')
                LOGGER.warning(jsde.doc)
    return build_info
