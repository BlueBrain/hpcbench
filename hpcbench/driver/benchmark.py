import contextlib
import copy
import glob
import json
import logging
import os
import re
import shlex
import uuid
from collections import namedtuple, Mapping
from os import path as osp

try:
    import magic
except ImportError:
    _HAS_MAGIC = False
else:
    _HAS_MAGIC = True

import six
import yaml
from cached_property import cached_property

from hpcbench.api import ExecutionContext, NoMetricException, Metric
from hpcbench.campaign import YAML_REPORT_FILE, JSON_METRICS_FILE
from .base import Enumerator, ClusterWrapper, write_yaml_report, Leaf
from .executor import Command
from hpcbench.toolbox.buildinfo import extract_build_info
from hpcbench.toolbox.collections_ext import nameddict
from hpcbench.toolbox.contextlib_ext import pushd
from hpcbench.toolbox.edsl import kwargsql
from hpcbench.toolbox.environment_modules import Module
from hpcbench.toolbox.functools_ext import listify
from hpcbench.toolbox.process import find_executable
from hpcbench.toolbox.spack import SpackCmd


class BenchmarkDriver(Enumerator):
    """Abstract representation of a benchmark, part of a campaign tag
    """

    def __init__(self, parent, benchmark, name, config):
        super(BenchmarkDriver, self).__init__(
            parent, benchmark.name, catch_child_exception=True
        )
        self.benchmark = benchmark
        self.config = BenchmarkDriver._prepare_config(config)
        self.name = name

    @classmethod
    def _prepare_config(cls, config):
        config = dict(config)
        config.setdefault('srun_options', [])
        if isinstance(config['srun_options'], six.string_types):
            config['srun_options'] = shlex.split(config['srun_options'])
        config['srun_options'] = [str(e) for e in config['srun_options']]
        return config

    @cached_property
    def children(self):
        categories = set()
        for execution in self.execution_matrix:
            categories.add(execution['category'])
        return categories

    def child_builder(self, child):
        return BenchmarkCategoryDriver(self, child)

    @cached_property
    def exec_context(self):
        tag = self.parent.name
        cluster = ClusterWrapper(self.root.network, tag, self.node)
        return ExecutionContext(
            cluster=cluster,
            logger=self.logger,
            node=self.node,
            srun_options=self.config['srun_options'],
            tag=tag,
            benchmark=self.name,
        )

    @property
    def execution_matrix(self):
        return self.benchmark.execution_matrix(self.exec_context)


class BenchmarkCategoryDriver(Enumerator):
    """Abstract representation of one benchmark to execute
    (one of "benchmarks" YAML tag values")"""

    ENVIRONMENT_VAR_TYPES = (int, float) + six.string_types

    def __init__(self, parent, category):
        super(BenchmarkCategoryDriver, self).__init__(parent, category)
        self.category = category
        self.benchmark = self.parent.benchmark
        self.config = parent.config
        self.exec_context = parent.exec_context

    @cached_property
    def commands(self):
        """Get all commands of the benchmark category

        :return generator of string
        """
        for child in self._children:
            with open(osp.join(child, YAML_REPORT_FILE)) as istr:
                command = yaml.safe_load(istr)['command']
                yield ' '.join(map(six.moves.shlex_quote, command))

    @property
    def _commands(self):
        exec_cls = self.root.execution_cls
        for em in self.parent.execution_matrix:
            for cmd in exec_cls.commands(self.campaign, self.config, em):
                yield cmd

    @cached_property
    @listify
    def children(self):
        for cmd in self._commands:
            valid = True
            category = cmd.execution.get('category')
            if category != self.category:
                continue
            # Override `environment` if specified in YAML
            if 'environment' in self.config:
                yaml_env = self.config.get('environment')
                if not yaml_env:
                    # empty dict or None wipes environment
                    cmd.execution['environment'] = {}
                else:
                    env = cmd.execution.setdefault('environment', {})
                    for name, value in yaml_env.items():
                        if value is None:
                            env.pop(name, None)
                        else:
                            # Have to exclude boolean manually because
                            # >>> isinstance(True, int)
                            # True
                            if not isinstance(value, bool) and isinstance(
                                value, self.ENVIRONMENT_VAR_TYPES
                            ):
                                env[name] = str(value)
                            else:
                                msg = 'Invalid type for environment variable "'
                                msg += name + '": "'
                                msg += value.__class__.__name__
                                msg += '". Valid types: '
                                msg += ', '.join(
                                    [t.__name__ for t in self.ENVIRONMENT_VAR_TYPES]
                                )
                                self.logger.error(msg)
                                valid = False
                    if not env:
                        self.config.pop('environment')
            # Override `modules` if specified in YAML
            if 'modules' in self.config:
                yaml_modules = self.config['modules'] or []
                cmd.execution['modules'] = list(yaml_modules)
            # Update spack config according to YAML
            # if None, then reset spack config
            if 'spack' in self.config:
                spack_c = self.config['spack']
                if spack_c is None:
                    cmd.execution['spack'] = {}
                else:
                    cmd.execution.setdefault('spack', {}).update(spack_c)
            # Enrich `metas` if specified in YAML
            if 'metas' in self.campaign:
                metas = dict(self.campaign.metas)
                metas.update(cmd.execution.setdefault('metas', {}))
                cmd.execution['metas'] = metas
            if valid:
                name = cmd.execution.get('name') or ''
                yield cmd, osp.join(name, self.child_id())

    def child_id(self):
        while True:
            child_id = str(uuid.uuid4()).split('-', 2)[0]
            if not hasattr(self, '__child_ids'):
                self.__child_ids = set()
            if child_id not in self.__child_ids:
                self.__child_ids.add(child_id)
                break
        return child_id

    def child_builder(self, child):
        del child  # unused

    @property
    def attempt_cls(self):
        config = self.parent.config.get('attempts')
        if config:
            fixed = config.get('fixed')
            if fixed is not None:
                assert isinstance(fixed, int)
                return FixedAttempts
            return DynamicAttempts
        return FixedAttempts

    @write_yaml_report
    @Enumerator.call_decorator
    def __call__(self, **kwargs):
        if "no_exec" not in kwargs:
            for run_dir in self._execute(**kwargs):
                yield run_dir
        else:
            self._extract_metrics(**kwargs)

    def _extract_metrics(self, **kwargs):
        runs = dict()
        for child in self.report['children']:
            child_yaml = osp.join(child, YAML_REPORT_FILE)
            with open(child_yaml) as istr:
                child_config = yaml.safe_load(istr)
            child_config.pop('children', None)
            runs.setdefault(self.category, []).append(child)
            with pushd(child):
                MetricsDriver(self, self.benchmark)(**kwargs)
        self.gather_metrics(runs)

    def _add_build_info(self, execution):
        executable = execution['command'][0]
        try:
            exepath = find_executable(executable, required=True)
        except NameError:
            self.logger.info(
                "Could not find exe %s to examine for build info", executable
            )
        else:
            try:
                file_type = magic.from_file(osp.realpath(exepath))
            except IOError as exc:
                self.logger.warn('Could not find file type of %s: %s', exepath, exc)
            else:
                if file_type.startswith('ELF'):
                    binfo = extract_build_info(exepath)
                    if binfo:
                        execution.setdefault('metas', {})['build_info'] = binfo
                else:
                    self.logger.info('%s is not pointing to an ELF executable', exepath)

    @contextlib.contextmanager
    def _module_env(self, execution):
        """Set current process environment according
        to execution `environment` and `modules`
        """
        env = copy.copy(os.environ)
        try:
            for mod in execution.get('modules') or []:
                Module.load(mod)
            os.environ.update(execution.get('environment') or {})
            yield
        finally:
            os.environ = env

    @contextlib.contextmanager
    def _spack_env(self, execution):
        env = copy.copy(os.environ)
        try:
            spack = SpackCmd()
            for spec in execution.get('spack', {}).get('specs', []):
                spack.install(spec)
                install_dir = spack.install_dir(spec)
                bin_dir = osp.join(install_dir, 'bin')
                if osp.exists(bin_dir):
                    path = os.environ.get('PATH', '')
                    path = bin_dir + os.pathsep + path
                    os.environ['PATH'] = path
            yield
        finally:
            os.environ = env

    def _execute(self, **kwargs):
        runs = dict()
        for command, run_dir in self.children:
            if _HAS_MAGIC and 'shell' not in command.execution:
                exc = command.execution
                with self._spack_env(exc), self._module_env(exc):
                    self._add_build_info(exc)
            else:
                self.logger.info(
                    "No build information recorded " "(libmagic available: %s)",
                    _HAS_MAGIC,
                )
            runs.setdefault(command.execution['category'], []).append(run_dir)
            with pushd(run_dir, mkdir=True):
                attempt = self.attempt_cls(self, command)
                for attempt in attempt(**kwargs):
                    pass
                yield run_dir
        self.gather_metrics(runs)

    def gather_metrics(self, runs):
        """Write a JSON file with the result of every runs
        """
        for run_dirs in runs.values():
            with open(JSON_METRICS_FILE, 'w') as ostr:
                ostr.write('[\n')
                for i in range(len(run_dirs)):
                    with open(osp.join(run_dirs[i], YAML_REPORT_FILE)) as istr:
                        data = yaml.safe_load(istr)
                        data.pop('category', None)
                        data.pop('command', None)
                        data['id'] = run_dirs[i]
                        json.dump(data, ostr, indent=2)
                    if i != len(run_dirs) - 1:
                        ostr.write(',')
                    ostr.write('\n')
                ostr.write(']\n')

    @cached_property
    def metrics(self):
        """Get content of the JSON metrics file
        """
        with open(JSON_METRICS_FILE) as istr:
            return json.load(istr)


class MetricsDriver(Leaf):
    """Abstract representation of metrics already
    built by a previous run
    """

    def __init__(self, parent, benchmark):
        super(MetricsDriver, self).__init__(parent)
        self.campaign = parent.campaign
        self.benchmark = benchmark

    @write_yaml_report
    @Enumerator.call_decorator
    def __call__(self, **kwargs):
        with open(YAML_REPORT_FILE) as istr:
            report = yaml.safe_load(istr)
        cat = report.get('category')
        metas = report.get('metas')
        all_extractors = self.benchmark.metrics_extractors
        if isinstance(all_extractors, Mapping):
            if cat not in all_extractors:
                raise Exception('No extractor for benchmark category %s' % cat)
            extractors = all_extractors[cat]
        else:
            extractors = all_extractors
        if not isinstance(extractors, list):
            extractors = [extractors]
        all_metrics = report.setdefault('metrics', [])
        for log in self.logs:
            metrics = {}
            for extractor in extractors:
                with extractor.context(log.path, log.log_prefix):
                    try:
                        run_metrics = extractor.extract(metas)
                    except NoMetricException:
                        pass
                    else:
                        MetricsDriver._check_metrics(extractor.metrics, run_metrics)
                        metrics.update(run_metrics)
            if metrics:
                rc = dict(context=log.context, measurement=metrics)
                all_metrics.append(rc)
        if self.benchmark.metric_required and not all_metrics:
            # at least one of the logs must provide metrics
            raise NoMetricException()
        return report

    class LocalLog(namedtuple('LocalLog', ['path', 'log_prefix'])):
        @property
        def context(self):
            return dict(executor='local')

    class SrunLog(namedtuple('SrunLog', ['path', 'log_prefix', 'node', 'rank'])):
        @property
        def context(self):
            return dict(executor='slurm', node=self.node, rank=self.rank)

    @property
    def logs(self):

        if self.report['executor'] == 'local':
            yield MetricsDriver.LocalLog(path=os.getcwd(), log_prefix='')
        else:
            STDOUT_RE_PATTERN = r'slurm-(\w+)-(\w+)\.stdout'
            STDOUT_RE = re.compile(STDOUT_RE_PATTERN)
            for file in glob.glob('slurm-*-*.stdout'):
                match = STDOUT_RE.match(file)
                if match:
                    node, rank = match.groups()
                    yield MetricsDriver.SrunLog(
                        path=os.getcwd(), log_prefix=file[:-6], node=node, rank=rank
                    )
                else:
                    logging.warn(
                        '"%s" does not match regular expression "%s"',
                        file,
                        STDOUT_RE_PATTERN,
                    )

    @classmethod
    def _check_metrics(cls, schema, metrics):
        """Ensure that returned metrics are properly exposed
        """
        for name, value in metrics.items():
            metric = schema.get(name)
            if not metric:
                message = "Unexpected metric '{}' returned".format(name)
                raise Exception(message)
            cls._check_metric(schema, metric, name, value)

    @classmethod
    def _check_metric(cls, schema, metric, name, value):
        if isinstance(metric, Metric):
            if not isinstance(value, metric.type):
                message = "Unexpected type for metrics {}:\n".format(name)
                message += "expected {}, but got {}".format(metric.type, type(value))
                raise Exception(message)
        elif isinstance(metric, list):
            if not isinstance(value, list):
                message = "Unexpected type for metrics {}".format(name)
                message += "expected {}, but got {}".format(list, type(value))
                raise Exception(message)
            for item in value:
                cls._check_metric(schema, metric[0], name, item)
        elif isinstance(metric, dict):
            if not isinstance(value, dict):
                message = "Unexpected type for metrics {}".format(name)
                message += "expected {}, but got {}".format(dict, type(value))
                raise Exception(message)
            cls._check_metrics(metric, value)
        else:
            message = "benchmark metric {} is neither ".format(metric)
            message += "a Metric, dict or list"
            raise Exception(message)


class FixedAttempts(Enumerator):
    def __init__(self, parent, command):
        super(FixedAttempts, self).__init__(parent)
        assert isinstance(command, Command)
        self.command = command
        self.paths = []
        self.config = parent.config
        self.exec_context = parent.exec_context

    __call__ = Enumerator._call_without_report

    @property
    def execution(self):
        return self.command.execution

    @property
    def benchmark(self):
        return self.parent.benchmark

    @cached_property
    def attempts_config(self):
        return self.parent.parent.config.get('attempts', {})

    @cached_property
    def attempts(self):
        return self.attempts_config.get('fixed', 1)

    @property
    def children(self):
        attempt = 1
        self.paths = []
        while self._should_run(attempt):
            path = 'attempt-' + str(attempt)
            self.paths.append(path)
            yield path
            attempt += 1
        attempt_path = self.last_attempt()
        for file_ in os.listdir(attempt_path):
            if file_ == YAML_REPORT_FILE:
                if osp.isfile(file_):
                    os.remove(file_)
            os.symlink(osp.join(attempt_path, file_), file_)

    def child_builder(self, child):
        def _wrap(**kwargs):
            driver = self.execution_layer()
            driver(**kwargs)
            if self.report['command_succeeded']:
                MetricsDriver(self, self.benchmark)(**kwargs)
            return self.report

        return _wrap

    def _should_run(self, attempt):
        return attempt <= self.attempts

    def execution_layer(self):
        """Build the proper execution layer
        """
        return self.root.execution_cls(self)

    def last_attempt(self):
        self._sort_attempts()
        return self.paths[-1]

    def _sort_attempts(self):
        if self.sort_config is not None:
            attempts = []
            for path in self.paths:
                with open(osp.join(path, YAML_REPORT_FILE)) as istr:
                    report = yaml.safe_load(istr)
                report['path'] = path
                attempts.append(report)
            attempts = sorted(attempts, **self.sort_config)
            self.paths = [report_['path'] for report_ in attempts]

    @cached_property
    def sort_config(self):
        config = self.attempts_config.get('sorted')
        if config is not None:
            sql = config.pop('sql', None)
            if sql is not None:
                if not isinstance(sql, list):
                    sql = [sql]

                def key_func(report):
                    return tuple([kwargsql.get(report, query) for query in sql])

                config['key'] = key_func
        return config


class DynamicAttempts(FixedAttempts):
    def __init__(self, parent, execution):
        super(DynamicAttempts, self).__init__(parent, execution)

    @cached_property
    def metric(self):
        return self.attempts_config['metric']

    @cached_property
    def attempts(self):
        return self.attempts_config['maximum']

    @cached_property
    def epsilon(self):
        return self.attempts_config.get('epsilon')

    @cached_property
    def percent(self):
        return self.attempts_config.get('percent')

    def _should_run(self, attempt):
        if attempt > self.attempts:
            return False
        if attempt < 3:
            return True
        with open(osp.join(self.paths[-2], YAML_REPORT_FILE)) as istr:
            data_n1 = nameddict(yaml.safe_load(istr))
        with open(osp.join(self.paths[-1], YAML_REPORT_FILE)) as istr:
            data_n = nameddict(yaml.safe_load(istr))
        return not self._metric_converged(data_n1, data_n)

    def _metric_converged(self, data_n1, data_n):
        def get_metric(report):
            # Only use first command result
            return report['metrics'][0]['measurement'][self.metric]

        value_n1 = get_metric(data_n1)
        value_n = get_metric(data_n)
        if self.epsilon is not None:
            return abs(value_n - value_n1) < self.epsilon
        assert self.percent is not None
        return abs(value_n - value_n1) < self.percent / 100.0
