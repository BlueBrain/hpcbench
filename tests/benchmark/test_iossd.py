import os
import os.path as osp
import unittest

from hpcbench.benchmark.iossd import (
    IOSSD,
    IOSSDExtractor,
)
from hpcbench.toolbox.contextlib_ext import (
    mkdtemp,
    pushd,
)
from . benchmark import AbstractBenchmarkTest


class TestIossd(AbstractBenchmarkTest, unittest.TestCase):
    EXPECTED_METRICS = {
        IOSSD.SSD_READ: dict(
            bandwidth=494.29915046691895,
        ),
        IOSSD.SSD_WRITE: dict(
            bandwidth=398.93171882629395,
        ),
    }

    def get_benchmark_clazz(self):
        return IOSSD

    def get_expected_metrics(self, category):
        return TestIossd.EXPECTED_METRICS[category]

    def get_benchmark_categories(self):
        return IOSSD.DEFAULT_CATEGORIES

    @property
    def attributes(self):
        return dict(
            executable='/path/to/fake'
        )

    def test_parse_bandwidth_linux(self):
        expected = 150
        self.assertEqual(
            IOSSDExtractor.parse_bandwidth_linux(expected * 1024, "KB/s"),
            expected
        )
        self.assertEqual(
            IOSSDExtractor.parse_bandwidth_linux(expected / 1024.0, "GB/s"),
            expected
        )
        self.assertEqual(
            IOSSDExtractor.parse_bandwidth_linux(expected * 1024 * 1024,
                                                 "bytes/s"),
            expected
        )
        with self.assertRaises(Exception):
            IOSSDExtractor.parse_bandwidth_linux(0, '?')

    def test_pre_execute_copy(self):
        with mkdtemp() as path:
            with pushd(path):
                benchmark = IOSSD()
                benchmark.pre_execute(None)
                self.assertTrue(osp.isfile(IOSSD.SCRIPT_NAME))
                self.assertTrue(os.access(IOSSD.SCRIPT_NAME, os.X_OK))
