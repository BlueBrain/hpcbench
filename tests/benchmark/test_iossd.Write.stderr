1024+0 records in
1024+0 records out
1073741824 bytes transferred in 2.071620 secs (418310226 bytes/sec)
